### PR TITLE
Add Sphinx Relay opt fragment

### PR DESCRIPTION
**Pull request by dennisreimann**
Integrates [Sphinx Relay](https://github.com/stakwork/sphinx-relay) – a Node.js wrapper around LND, handling connectivity and storage for [Sphinx Chat](https://sphinx.chat/). 
This will be a handy addition for e.g. podcasters, that can use this integration to host their own [tribe](https://tribes.sphinx.chat/) and earn sats by streaming their podcast.

Leverages their official Docker image and connects it with our LND. The `user: "0:0"` config is required, because their image runs with a separate user `node`, which cannot access the linked LND files – that's why we set the user to `root` as in our other containers. 

Sphinx uses keysend, that's why we need to require the opt as well.

## Testing

To test this PR while it isn't merged, you can deploy it as a [custom fragment](https://docs.btcpayserver.org/Docker/#how-can-i-customize-the-generated-docker-compose-file):

- SSH into your BTCPay Server and go to the main BTCPay Server directory (where the `btcpay-setup.sh` script is located)
- create the file `docker-compose-generator/docker-fragments/opt-add-sphinxrelay.custom.yml` with the content of `opt-add-sphinxrelay.yml`
- add the diff snippet from `Production/nginx.tmpl` to the `Production/nginx.tmpl` file on your server
- run the following commands:

```bash
export BTCPAYGEN_ADDITIONAL_FRAGMENTS="$BTCPAYGEN_ADDITIONAL_FRAGMENTS;opt-add-sphinxrelay.custom"
. ./btcpay-setup.sh -i
````

Once the setup is done you can check the Sphinx log using `docker logs -f generated_sphinxrelay_1`.

You should see a log like this, which also give you a QR code to connect.

sphinx

### CODE REVIEW

**Comment by taxmeifyoucan** [comment]
Successfully deployed Sphinx in my Btcpay docker! Two issues I ran into so far, not sure if I should open issue in Sphinx repo, just to be sure reporting here. 

1. Opening logs after initial deployment, it is successfully running but there is no QR code. After restarting service with `docker restart generated_sphinxrelay_1`, QR code appeared as expected
2. Can't join tribes. After loading tribe info and hitting 'Join Group' button, nothing happens for a while, then I am back to home screen and tribe is not there. Creating own tribe wasn't problem. Here are logs of what happen on Sphinx Relay when I try to join tribe: 
```
=> sending to 1 contacts
[lightning] [ 'keysendMessage' ]
[lightning] [ 'keysend' ]
(node:1) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, open '/relay/lnd/router.macaroon'
    at Object.openSync (fs.js:462:3)
    at Object.readFileSync (fs.js:364:35)
    at Object.getMacaroon (/relay/dist/src/utils/macaroon.js:17:22)
    at loadCredentials (/relay/dist/src/utils/lightning.js:37:31)
    at loadRouter (/relay/dist/src/utils/lightning.js:334:31)
    at /relay/dist/src/utils/lightning.js:298:38
    at Generator.next (<anonymous>)
    at /relay/dist/src/utils/lightning.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/relay/dist/src/utils/lightning.js:4:12)
    at /relay/dist/src/utils/lightning.js:250:16
    at new Promise (<anonymous>)
    at keysend (/relay/dist/src/utils/lightning.js:249:12)
    at /relay/dist/src/utils/lightning.js:356:43
    at Generator.next (<anonymous>)
    at /relay/dist/src/utils/lightning.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/relay/dist/src/utils/lightning.js:4:12)
    at /relay/dist/src/utils/lightning.js:350:20
    at new Promise (<anonymous>)
    at Object.<anonymous> (/relay/dist/src/utils/lightning.js:349:16)
    at Generator.next (<anonymous>)
    at /relay/dist/src/utils/lightning.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/relay/dist/src/utils/lightning.js:4:12)
    at Object.keysendMessage (/relay/dist/src/utils/lightning.js:347:12)
    at /relay/dist/src/network/send.js:182:31
    at Generator.next (<anonymous>)
    at fulfilled (/relay/dist/src/network/send.js:5:58)
(node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
```

**Comment by dennisreimann** [comment]
Cool, thanks for testing it! I think for 1. we need to get in touch with the Sphinx team. 

For 2. I'll consult with @rockstardev on how to best provide the signer and router macaroon to Sphinx as well. These aren't in the lnd data dir like the admin.macaroon, but Sphinx expects them to be in the same directory. (They are currently nested in `data/chain/bitcoin/mainnet/`)

**Comment by dennisreimann** [comment]
Once stakwork/sphinx-relay#181 gets merged we can continue here :)

I'll verify the QR code not being shown initially issue and will bring it up with them in case it still doesn't work with the fix applied.

**Comment by dennisreimann** [comment]
Alright, the Sphinx Relay PR got merged and with the updates I just pushed I can confirm it works for me.

You can give the updated version another try 🚀 @taxmeifyoucan @Eskyee

**Comment by Eskyee** [comment]
Will test this afternoon!! Cheers

File: docker-compose-generator/docker-fragments/opt-add-sphinxrelay.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      - "10009"
+  sphinxrelay:
+    image: sphinxlightning/sphinx-relay:v2.0.12
+    user: "0:0"
+    restart: unless-stopped
+    expose:
+      - "3300"
+    volumes:
+      - "sphinxrelay_datadir:/relay/.lnd"
+      - "lnd_bitcoin_datadir:/relay/lnd:ro"
```
Comment (NicolasDorier) [inline docker-compose-generator/docker-fragments/opt-add-sphinxrelay.yml:19]: I wonder if you could share only the macaroon and not all the folder.

Comment (NicolasDorier) [reply docker-compose-generator/docker-fragments/opt-add-sphinxrelay.yml:19]: Tried to search google, does not seem possible

Comment (dennisreimann) [reply docker-compose-generator/docker-fragments/opt-add-sphinxrelay.yml:19]: I've tried mounting single files and also didn't find anything. That's also why I needed the PR in Sphinx Relay itself – my initial idea was mounting the macaroons to the locations Sphinx expects, but that didn't work.

**Comment by taxmeifyoucan** [comment]
Awesome! I tested it on new Btcpay docker instance and I still experience the issue with QR code not showing immediately after install - container has to be restarted. Here are logs after the installation: 
```
=> env: production
=> [db] starting setup...
=> [db] sync now
=> [db] done syncing
=> [db] setup done
=> 2021-04-23 09:58:47 [lnd] reconnecting... attempt #1
Node listening on 3300.
(node:1) DeprecationWarning: grpc.load: Use the @grpc/proto-loader module with grpc.loadPackageDefinition instead
=> [lnd] connected! 2021-04-23 09:58:50
>> FINISH SETUP
=> Relay version: , commit: 
>> aXA6Omh0dHBzOi8vYnRjcGF5LmxpZ2h0bmluZ2NvbmR1Y3RvcnMubmV0L3NwaGlueHJlbGF5Lzo6Y2Y5MzYxOGM1ZTA5ZjZlNjFmMjcxNTM2NGExMTUzM2Y=
[lightning] [ 'signBuffer' ]
[tribes] try to connect: tls://tribes.sphinx.chat:8883
[db] created node owner contact, id: 1
[tribes] connected!
[tribes] subscribed! 02347e64c80e30eb3f4d6d1b0d19a5e91368973029e62347e9810380715080f53e/#
```
Tribes work, not problem anymore! And everything else too, didn't find any new issue.

**Comment by dennisreimann** [comment]
The missing QR code is also described in stakwork/sphinx-relay#183.

The QR code is an encoded version of the connection string, which is displayed below the `=> Relay version: , commit:` line.
You can copy it and paste it in the app, it will work the same as the QR code.

I think we can merge this then as @Eskyee also confirmed it works for him.

**Comment by Eskyee** [comment]
Yes I can confirm.. good job 👍

**Review by Eskyee — APPROVED** [review]

**Comment by rockstardev** [comment]
Tested the PR during last 10 minutes of stream @dennisreimann, everything went as expected it seems. Missing display of QR code (as @lopp described), but there is connection string.

You can see how it all went here: https://youtu.be/5bho645cMBo?t=3436

**Comment by dennisreimann** [comment]
@rockstardev Awesome, thanks for reviewing and covering this! 💚

**Comment by jimbojw** [comment]
I'm currently setting up sphinx-relay for the first time against my BTCPayServer/lnd. Should I use this procedure?

**Comment by dennisreimann** [comment]
@jimbojw as the opt fragment isn't part of an official release yet, you can test and review it using the manual install method described above. Let me know if you have more questions.

**Comment by jimbojw** [comment]
What needs to happen for this to be merged and released? Trying to gauge whether to manually test or wait for official.

**Comment by jimbojw** [comment]
Never mind, I used the test instructions, YOLO!

**Comment by dennisreimann** [comment]
We received some good feedback already, further review and testing will certainly help getting this merged.

**Comment by jimbojw** [comment]
I had the same problem as @taxmeifyoucan. The QR code did not show up immediately, but did after I restarted `generated_sphinxrelay_1`. But connected now.

**Review by Kukks — APPROVED** [review]
LGTM, is anything missing to merge?

**Comment by dennisreimann** [comment]
Nope, it's ready and we have three confirmations of a working install.

**Comment by jimbojw** [comment]
I used these instructions to set up my Sphinx relay, and now I need to update it. What are the instructions to update to sphinx-relay v2.1.0?

**Comment by dennisreimann** [comment]
@jimbojw we need to update it here in the repo. Feel free to submit a pull request with an update. See the files modified in this PR for reference on what to update where.

**Pull request merged**